### ISSUE TITLE
Calculation of epochs vs steps and total progress is wrong

### ISSUE THREAD

**Issue by a-l-e-x-d-s-9** [bghira/SimpleTuner#509]
In my training, I have 6183 images per epoch, batch size 27 - i.e. 229 batches per epoch = 229 steps per epoch, GRADIENT_ACCUMULATION_STEPS=1.
I have been running training with a target of 100 epochs, which is translated to 22900 total steps.
I see that the epoch number is wrong. For 7800 steps it reports 18 epochs, in terminal and in wandb report. 7800 steps are 34% - which is correct and should be corresponding to epoch number 35, not 18.
The epochs are lagging behind throughout the whole training, I show 7800 steps as an example.
Here is what I see in the terminal: 
`Epoch 18/100, Steps:  34%|█████████▌                  | 7800/22900`

**Comment by bghira** [comment]
yeah its never going to be correct unfortunately. measuring epochs isnt trivial at all.

**Comment by a-l-e-x-d-s-9** [comment]
But we know the exact number of batches per epoch and steps per epoch, from the start. 
Doesn't the script need to reshuffle images into batches before each epoch, and with the wrong calculation of epochs, it uses the old order?
What am I missing that makes the calculation or detection of epochs complicated?

**Comment by bghira** [comment]
oh, you probably have 'repeats' set to 1?

**Comment by bghira** [comment]
please be sure to support your issue reports with config details and ideally logs related to the generation of buckets and the batch size.

**Comment by a-l-e-x-d-s-9** [comment]
The repeats are set to 1, and the dataset has probability=1, crop=false.
The info about images split between aspect ratio buckets doesn't appear in wandb log, and the server has already been wiped out.
I guess that if you make random cropping, and use probability not equal to 1, you can get different aspect ratio buckets at the start of each epoch, potentially changing a number of batches and steps, but in this case, all the info about steps and batches is known to the script, and it can be saved to a variable that counts steps per current epoch, to make a correct update of epoch count. Isn't there a point in the code where you are preparing images for the next epoch with cropping and aspect ratio buckets and split to batches - can't at this point the counter of epoch be advanced and represent the correct epoch number?
If the cropping and probability, can change the number of batches per epoch, is the total amount of steps estimated at start?

wandb log:
 ``` log
1 2024-06-20 17:48:06,590 [INFO] (__main__) 
 2 ***** Running training *****
 3 -  Num batches = 229
 4 -  Num Epochs = 100
 5   - Current Epoch = 3
 6 -  Total train batch size (w. parallel, distributed & accumulation) = 27
 7   - Instantaneous batch size per device = 27
 8   - Gradient Accumulation steps = 1
 9 -  Total optimization steps = 22900
10   - Steps completed: 1080
11 -  Total optimization steps remaining = 21820
12 Epoch 5/100, Steps:  10%|██▉                          | 2280/22900 [2:16:47<25:56:03,  4.53s/it, lr=1.5e-6, step_loss=0.0978]Model weights saved in /workspace/output/temporary-2280/transformer/diffusion_pytorch_model.safetensors
13 Epoch 3/100, Steps:   5%|█▋                              | 1200/22900 [13:27<40:36:13,  6.74s/it, lr=1.5e-6, step_loss=0.107]
14 Epoch 3/100, Steps:   6%|█▊                              | 1320/22900 [27:31<40:37:16,  6.78s/it, lr=1.5e-6, step_loss=0.116]
15 Epoch 4/100, Steps:   6%|██                               | 1440/22900 [41:02<41:39:22,  6.99s/it, lr=1.5e-6, step_loss=0.11]
16 Epoch 4/100, Steps:   7%|██▏                             | 1560/22900 [54:58<35:19:09,  5.96s/it, lr=1.5e-6, step_loss=0.115]
17 Epoch 4/100, Steps:   7%|██▏                           | 1680/22900 [1:08:38<43:04:26,  7.31s/it, lr=1.5e-6, step_loss=0.115]
18 Epoch 4/100, Steps:   8%|██▎                           | 1800/22900 [1:22:31<42:14:57,  7.21s/it, lr=1.5e-6, step_loss=0.113]
19 Epoch 5/100, Steps:   8%|██▌                            | 1920/22900 [1:36:07<38:59:09,  6.69s/it, lr=1.5e-6, step_loss=0.13]
20 Epoch 5/100, Steps:   9%|██▋                           | 2040/22900 [1:49:57<38:05:50,  6.57s/it, lr=1.5e-6, step_loss=0.126]
21 Epoch 5/100, Steps:   9%|██▊                           | 2160/22900 [2:03:39<41:19:19,  7.17s/it, lr=1.5e-6, step_loss=0.121]
22 Epoch 18/100, Steps:  34%|█████████▌                  | 7800/22900 [12:55:35<31:49:22,  7.59s/it, lr=1.5e-6, step_loss=0.123]Model weights saved in /workspace/output/temporary-7800/transformer/diffusion_pytorch_model.safetensors
23 Epoch 6/100, Steps:  10%|███                          | 2400/22900 [2:31:08<36:22:40,  6.39s/it, lr=1.5e-6, step_loss=0.0999]
24 Epoch 6/100, Steps:  11%|███▎                          | 2520/22900 [2:44:18<27:27:45,  4.85s/it, lr=1.5e-6, step_loss=0.119]
25 Epoch 6/100, Steps:  12%|███▌                           | 2640/22900 [2:59:02<43:51:05,  7.79s/it, lr=1.5e-6, step_loss=0.11]
26 Epoch 7/100, Steps:  12%|███▌                          | 2760/22900 [3:12:26<43:27:31,  7.77s/it, lr=1.5e-6, step_loss=0.114]
27 Epoch 7/100, Steps:  13%|███▋                         | 2880/22900 [3:26:49<38:03:28,  6.84s/it, lr=1.5e-6, step_loss=0.0854]
28 Epoch 7/100, Steps:  13%|███▊                          | 2930/22900 [3:32:55<41:11:52,  7.43s/it, lr=1.5e-6, step_loss=0.132]
29 Epoch 7/100, Steps:  13%|███▉                          | 3000/22900 [3:40:35<40:51:37,  7.39s/it, lr=1.5e-6, step_loss=0.105]
30 Epoch 7/100, Steps:  14%|████                          | 3120/22900 [3:55:02<37:44:13,  6.87s/it, lr=1.5e-6, step_loss=0.133]
31 Epoch 8/100, Steps:  14%|████▏                         | 3240/22900 [4:08:19<37:45:05,  6.91s/it, lr=1.5e-6, step_loss=0.111]
32 Epoch 8/100, Steps:  15%|████▍                         | 3360/22900 [4:23:01<35:37:05,  6.56s/it, lr=1.5e-6, step_loss=0.147]
33 Epoch 8/100, Steps:  15%|████▍                        | 3480/22900 [4:36:28<39:09:51,  7.26s/it, lr=1.5e-6, step_loss=0.0982]
34 Epoch 8/100, Steps:  16%|████▋                         | 3600/22900 [4:50:34<34:50:20,  6.50s/it, lr=1.5e-6, step_loss=0.113]
35 Epoch 9/100, Steps:  16%|████▋                        | 3720/22900 [5:04:02<36:10:55,  6.79s/it, lr=1.5e-6, step_loss=0.0995]
36 Epoch 9/100, Steps:  17%|█████                         | 3840/22900 [5:18:21<36:30:07,  6.89s/it, lr=1.5e-6, step_loss=0.106]
37 Epoch 9/100, Steps:  17%|█████                        | 3960/22900 [5:32:08<37:15:43,  7.08s/it, lr=1.5e-6, step_loss=0.0936]
38 Epoch 9/100, Steps:  18%|█████▎                        | 4080/22900 [5:46:10<32:59:34,  6.31s/it, lr=1.5e-6, step_loss=0.109]
39 Epoch 10/100, Steps:  18%|█████▏                      | 4200/22900 [6:00:21<38:39:15,  7.44s/it, lr=1.5e-6, step_loss=0.0997]
40 Epoch 10/100, Steps:  19%|█████▎                      | 4320/22900 [6:14:10<32:23:41,  6.28s/it, lr=1.5e-6, step_loss=0.0935]
41 Epoch 10/100, Steps:  19%|█████▌                       | 4440/22900 [6:27:47<34:22:47,  6.70s/it, lr=1.5e-6, step_loss=0.106]
42 Epoch 10/100, Steps:  20%|█████▊                       | 4560/22900 [6:41:36<34:35:22,  6.79s/it, lr=1.5e-6, step_loss=0.122]
43 Epoch 11/100, Steps:  20%|█████▉                       | 4647/22900 [6:51:32<35:20:54,  6.97s/it, lr=1.5e-6, step_loss=0.114]
44 Epoch 11/100, Steps:  20%|█████▉                       | 4680/22900 [6:56:13<39:34:26,  7.82s/it, lr=1.5e-6, step_loss=0.157]
45 Epoch 11/100, Steps:  21%|██████                       | 4800/22900 [7:10:02<22:45:32,  4.53s/it, lr=1.5e-6, step_loss=0.131]
46 Epoch 11/100, Steps:  21%|██████▏                      | 4920/22900 [7:24:44<32:26:56,  6.50s/it, lr=1.5e-6, step_loss=0.109]
47 Epoch 12/100, Steps:  22%|██████▍                      | 5040/22900 [7:37:48<27:18:55,  5.51s/it, lr=1.5e-6, step_loss=0.147]
48 Epoch 12/100, Steps:  23%|██████▌                      | 5160/22900 [7:52:26<30:51:28,  6.26s/it, lr=1.5e-6, step_loss=0.098]
49 Epoch 12/100, Steps:  23%|██████▍                     | 5280/22900 [8:05:36<36:44:21,  7.51s/it, lr=1.5e-6, step_loss=0.0903]
50 Epoch 12/100, Steps:  24%|██████▊                      | 5400/22900 [8:19:56<35:49:39,  7.37s/it, lr=1.5e-6, step_loss=0.105]
51 Epoch 13/100, Steps:  24%|██████▉                      | 5520/22900 [8:33:23<36:19:57,  7.53s/it, lr=1.5e-6, step_loss=0.108]
52 Epoch 13/100, Steps:  25%|███████▏                     | 5640/22900 [8:47:40<33:06:59,  6.91s/it, lr=1.5e-6, step_loss=0.112]
53 Epoch 13/100, Steps:  25%|███████▎                     | 5760/22900 [9:00:59<34:43:29,  7.29s/it, lr=1.5e-6, step_loss=0.107]
54 Epoch 13/100, Steps:  26%|███████▍                     | 5880/22900 [9:15:12<32:24:14,  6.85s/it, lr=1.5e-6, step_loss=0.124]
55 Epoch 14/100, Steps:  26%|███████▎                    | 6000/22900 [9:29:10<34:29:45,  7.35s/it, lr=1.5e-6, step_loss=0.0948]
56 Epoch 14/100, Steps:  27%|███████▍                    | 6120/22900 [9:43:31<30:38:12,  6.57s/it, lr=1.5e-6, step_loss=0.0967]
57 Epoch 14/100, Steps:  27%|███████▉                     | 6240/22900 [9:56:56<32:20:48,  6.99s/it, lr=1.5e-6, step_loss=0.101]
58 Epoch 14/100, Steps:  28%|███████▊                    | 6360/22900 [10:10:57<33:14:49,  7.24s/it, lr=1.5e-6, step_loss=0.109]
59 Epoch 15/100, Steps:  28%|███████▉                    | 6480/22900 [10:24:32<35:34:35,  7.80s/it, lr=1.5e-6, step_loss=0.108]
60 Epoch 15/100, Steps:  29%|████████▎                    | 6600/22900 [10:38:59<30:52:29,  6.82s/it, lr=1.5e-6, step_loss=0.14]
61 Epoch 15/100, Steps:  29%|████████▏                   | 6720/22900 [10:52:40<34:24:38,  7.66s/it, lr=1.5e-6, step_loss=0.146]
62 Epoch 15/100, Steps:  30%|████████                   | 6840/22900 [11:06:25<28:46:28,  6.45s/it, lr=1.5e-6, step_loss=0.0963]
63 Epoch 16/100, Steps:  30%|████████▌                   | 6960/22900 [11:20:07<30:43:20,  6.94s/it, lr=1.5e-6, step_loss=0.111]
64 Epoch 16/100, Steps:  31%|████████▋                   | 7080/22900 [11:34:11<26:09:48,  5.95s/it, lr=1.5e-6, step_loss=0.104]
65 Epoch 16/100, Steps:  31%|█████████                    | 7200/22900 [11:47:55<31:41:46,  7.27s/it, lr=1.5e-6, step_loss=0.11]
66 Epoch 16/100, Steps:  32%|████████▋                  | 7320/22900 [12:01:18<18:27:09,  4.26s/it, lr=1.5e-6, step_loss=0.0723]
67 Epoch 17/100, Steps:  32%|█████████                   | 7440/22900 [12:15:21<28:25:30,  6.62s/it, lr=1.5e-6, step_loss=0.132]
68 Epoch 17/100, Steps:  33%|█████████▏                  | 7560/22900 [12:28:22<25:59:15,  6.10s/it, lr=1.5e-6, step_loss=0.141]
69 Epoch 17/100, Steps:  34%|█████████▍                  | 7680/22900 [12:42:44<27:34:00,  6.52s/it, lr=1.5e-6, step_loss=0.128]
70 Epoch 18/100, Steps:  34%|█████████▌                  | 7800/22900 [12:55:59<29:03:40,  6.93s/it, lr=1.5e-6, step_loss=0.123]
```

multidatabackend.json
``` jsin
[
    {
        "id": "all_dataset",
        "type": "local",
        "instance_data_dir": "/workspace/input/dataset",
        "crop": false,
        "crop_style": "random",
        "crop_aspect": "preserve",
        "resolution": 0.5,
        "resolution_type": "area",
        "minimum_image_size": 0.1,
        "maximum_image_size": 1.0,
        "target_downsample_size":0.55,
        "prepend_instance_prompt": false,
        "instance_prompt": null,
        "only_instance_prompt": false,
        "caption_strategy": "textfile",
        "cache_dir_vae": "/workspace/cache_images/",
        "vae_cache_clear_each_epoch": false,
        "probability": 1.0,
        "repeats": 1,
        "text_embeds": "alt-embed-cache",
        "skip_file_discovery": "",
        "preserve_data_backend_cache": true
    },
    {
        "id": "alt-embed-cache",
        "dataset_type": "text_embeds",
        "default": true,
        "type": "local",
        "cache_dir": "/workspace/cache_text_embeds/"
    }
]
```

**Comment by bghira** [comment]
repeats=1 actually doubles the sample which makes it the 17th epoch you are on

**Comment by a-l-e-x-d-s-9** [comment]
In a report here: https://github.com/bghira/SimpleTuner/issues/484
When using repeat 5 with a single dataset it didn't do anything.
You said that repeats are supposed to only work with two or more datasets.

**Comment by bghira** [comment]
i said they're intended for use with >1 dataset. and if they're not working with 1 dataset, that it might be a bug, but that it also wouldn't matter for most purposes other than learning rate schedule concerns

**Comment by bghira** [comment]
for the simplest example, it is correct:

```
2024-06-21 09:29:06,714 [INFO] (MultiAspectSampler-single_image_dataset) 
(Rank: 0)     -> Number of seen images: 0
(Rank: 0)     -> Number of unseen images: 1
(Rank: 0)     -> Current Bucket: None
(Rank: 0)     -> 1 Buckets: ['1.0']
(Rank: 0)     -> 0 Exhausted Buckets: []
2024-06-21 09:29:07,461 [INFO] (__main__) 
***** Running training *****
-  Num batches = 1
-  Num Epochs = 20000
  - Current Epoch = 1
-  Total train batch size (w. parallel, distributed & accumulation) = 1
  - Instantaneous batch size per device = 1
  - Gradient Accumulation steps = 1
-  Total optimization steps = 20000
-  Total optimization steps remaining = 20000
```

**Comment by bghira** [comment]
```
Epoch 100/20000, Steps:   1%|▎                                   | 200/20000 [00:03<04:30, 73.20it/s, lr=0, step_loss=0.0403]
```

when repeats=1

**Comment by bghira** [comment]
```
Epoch 200/20000, Steps:   1%|▎                                    | 199/20000 [00:03<05:28, 60.25it/s, lr=0, step_loss=0.169]
```

repeats=0 / undeclared

**Comment by bghira** [comment]
ah. the length of the dataset wasn't being multiplied by the repeat count.

however, the epoch tracking was still being done correctly. this really only changes the total length of the training run to resolve it:

```
***** Running training *****
-  Num batches = 2
-  Num Epochs = 20000
  - Current Epoch = 1
-  Total train batch size (w. parallel, distributed & accumulation) = 1
  - Instantaneous batch size per device = 1
  - Gradient Accumulation steps = 1
-  Total optimization steps = 40000
-  Total optimization steps remaining = 40000
Epoch 100/20000, Steps:   0%|▏                                    | 199/40000 [00:03<08:30, 77.89it/s, lr=0, step_loss=0.169]
```

note that num batches is now 2 for a single image dataset with repeats=1.

epochs=20000 but total steps=40000 because batch size is 1 and we have double the sample steps to take on the repeat of 1.

**Comment by a-l-e-x-d-s-9** [comment]
Ok,
With repeats=1 and 20,000 epochs, the total number of steps should be (20,000 x (1 + 1)) = 40,000.
It makes sense.
Although, your usage of "repeats" is different from kohya. In kohya the meaning of repeats is how many times (multiplication) a folder will be used per epoch, and you are using it in a meaning of how many **additional times** it will be added per epoch. It's fine if you don't want to be consistent in terminology with other repositories, but it will be confusing to people coming from kohya.

**Comment by bghira** [comment]
i mean, kohya's is the confusing one, to me - repeats is how many times it is repeated. 0 would be no repeats, just a single pass?

**Comment by a-l-e-x-d-s-9** [comment]
Yes, you are correct in the colloquial sense of repeats.
But using "repeats" the way kohya uses it, provides the benefit of using 0, so you don't need to remove the information from the file when you don't want to use it in a specific training, you just use repeats=0 to disable it.
In SimpleTuner - using probability=0, will disable the dataset?

**Comment by bghira** [comment]
yes.

but you should just set `disable=true`
